src/bluetooth.rs

```diff
@@ -6,187 +6,210 @@
+    Android(Arc<BluetoothGATTCharacteristicAndroid>),
+    #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"), all(target_os = "android", feature = "bluetooth"))))]
+    Empty(Arc<BluetoothGATTCharacteristicEmpty>),
     #[cfg(feature = "bluetooth-test")]
-    gatt_characteristic: Arc<FakeBluetoothGATTCharacteristic>,
+    Mock(Arc<FakeBluetoothGATTCharacteristic>),
 }
 
 #[derive(Clone, Debug)]
-pub struct BluetoothGATTDescriptor {
-    characteristic: RefCell<BluetoothGATTCharacteristic>,
+pub enum BluetoothGATTDescriptor {
     #[cfg(all(target_os = "linux", feature = "bluetooth"))]
-    gatt_descriptor: Arc<BluetoothGATTDescriptorBluez>,
+    Bluez(Arc<BluetoothGATTDescriptorBluez>),
     #[cfg(all(target_os = "android", feature = "bluetooth"))]
-    gatt_descriptor: Arc<BluetoothGATTDescriptorAndroid>,
-    #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"), all(target_os = "android", feature = "bluetooth"), feature = "bluetooth-test")))]
-    gatt_descriptor: Arc<BluetoothGATTDescriptorEmpty>,
+    Android(Arc<BluetoothGATTDescriptorAndroid>),
+    #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"), all(target_os = "android", feature = "bluetooth"))))]
+    Empty(Arc<BluetoothGATTDescriptorEmpty>),
     #[cfg(feature = "bluetooth-test")]
-    gatt_descriptor: Arc<FakeBluetoothGATTDescriptor>,
+    Mock(Arc<FakeBluetoothGATTDescriptor>),
+}
+
+macro_rules! get_inner_and_call(
+    ($enum_value: expr, $enum_type: ident, $function_name: ident) => {
+        match $enum_value {
+            #[cfg(all(target_os = "linux", feature = "bluetooth"))]
+            &$enum_type::Bluez(ref bluez) => bluez.$function_name(),
+            #[cfg(all(target_os = "android", feature = "bluetooth"))]
+            &$enum_type::Android(ref android) => android.$function_name(),
+            #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"), all(target_os = "android", feature = "bluetooth"))))]
+            &$enum_type::Empty(ref empty) => empty.$function_name(),
+            #[cfg(feature = "bluetooth-test")]
+            &$enum_type::Mock(ref fake) => fake.$function_name(),
+        }
+    };
+
+    ($enum_value: expr, $enum_type: ident, $function_name: ident, $value: expr) => {
+        match $enum_value {
+            #[cfg(all(target_os = "linux", feature = "bluetooth"))]
+            &$enum_type::Bluez(ref bluez) => bluez.$function_name($value),
+            #[cfg(all(target_os = "android", feature = "bluetooth"))]
+            &$enum_type::Android(ref android) => android.$function_name($value),
+            #[cfg(not(any(all(target_os = "linux", feature = "bluetooth"), all(target_os = "android", feature = "bluetooth"))))]
+            &$enum_type::Empty(ref empty) => empty.$function_name($value),
+            #[cfg(feature = "bluetooth-test")]
+            &$enum_type::Mock(ref fake) => fake.$function_name($value),
+        }
+    };
+);
+
+macro_rules! get_inner_and_call_test_func {
```

Comment:
If this only used in `#[cfg(feature = "bluetooth-test")]` functions, this should be guarded as well.